presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFilterFunction.java

```diff
@@ -16,17 +16,36 @@
 import com.facebook.presto.common.block.Block;
 import com.facebook.presto.common.block.BlockBuilder;
 import com.facebook.presto.common.type.Type;
+import com.facebook.presto.spi.function.ComplexTypeFunctionDescriptor;
 import com.facebook.presto.spi.function.Description;
 import com.facebook.presto.spi.function.ScalarFunction;
+import com.facebook.presto.spi.function.ScalarFunctionDescriptor;
+import com.facebook.presto.spi.function.ScalarFunctionLambdaArgumentDescriptor;
+import com.facebook.presto.spi.function.ScalarFunctionLambdaDescriptor;
 import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.function.StaticMethodPointer;
 import com.facebook.presto.spi.function.TypeParameter;
 import com.facebook.presto.spi.function.TypeParameterSpecialization;
 import io.airlift.slice.Slice;
 
 import static java.lang.Boolean.TRUE;
 
 @Description("return array containing elements that match the given predicate")
-@ScalarFunction(value = "filter", deterministic = false)
+@ScalarFunction(value = "filter", deterministic = false, descriptor = @ScalarFunctionDescriptor(
+        isAccessingInputValues = true,
+        isTransformingInput = false,
+        isReturningInputValues = true,
+        argumentIndicesContainingMapOrArray = {0},
+        outputToInputTransformationFunction = {},
+        lambdaDescriptors = {
+                @ScalarFunctionLambdaDescriptor(
+                        callArgumentIndex = 1,
+                        lambdaArgumentDescriptors = {
+                                @ScalarFunctionLambdaArgumentDescriptor(
+                                        lambdaArgumentIndex = 0,
+                                        callArgumentIndex = 0,
+                                        lambdaArgumentToInputTransformationFunction = @StaticMethodPointer(
```

Comment:
In general, we should make annotations simple, because they're in SPI. They should not expose any internal details, and not give too many options to the developer integrating using SPI.

Reply:
There is a trade off between flexibility and readability/simplicity. The ability to define a function descriptor for external developers was essential. Therefore, we have such complexity. Function descriptor essentially tells us about what function and its lambdas are doing with the input.